### PR TITLE
Laboratory work #2, Andrey Klimov - 23FPL1

### PR DESCRIPTION
**Pull request by QuietHellsPage**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,6 +23,12 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+    for item in text:
+        if not item.isalpha() and item != ' ':
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:29]: use continue if item is alpha or whitespace

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +44,10 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not isinstance(stopwords, list) or not all(isinstance(symbol, str) for symbol in tokens) or not all(isinstance(symb, str) for symb in stopwords) or not tokens or not stopwords:
+        return None
+    filtered_tokens = [word for word in tokens if word not in stopwords]
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:49]: great!

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -62,6 +88,13 @@ def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, floa
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(vocab, list) or not all(isinstance(item, str) for item in vocab) or not vocab or not isinstance(document_tokens, list) or not all(isinstance(item, str) for item in document_tokens) or not document_tokens:
+        return None
+    tf_result = {}
+    union = set(vocab).union(set(document_tokens))
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:94]: great!

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -77,6 +110,16 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
...
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(vocab, list) or not all(isinstance(item, str) for item in vocab) or not isinstance(documents, list) or not all(isinstance(docum, list) for docum in documents) or not all(isinstance(item, str) for docum in documents for item in docum) or not vocab or not documents:
+        return None
+    idf_result = {}
+    for word in vocab:
+        words_count = 0
+        for doc in documents:
+            if word in doc:
+                words_count += 1
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:120]: Do you think it is possible to use sum() here to calculate words_count? If so, try to implement it

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -92,6 +135,12 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tf, dict) or not isinstance(idf, dict) or not all(isinstance(key, str) for key in tf) or not all(isinstance(key, str) for key in idf) or not all(isinstance(value, float) for value in tf.values()) or not all(isinstance(value, float) for value in idf.values()) or not tf or not idf:
+        return None
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:139]: try to use dict comprehension

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great job!!!

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:29]: Done, but can you explain why did I have to do this? I thought, that my previous code was normal. Maybe some new checks were needed and that's why I had to make new func?

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:120]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:146]: Done

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -3,6 +3,11 @@
 
 Text retrieval with BM25
 """
+
+
+from docutils.nodes import document
+from sphinx.cmd.quickstart import nonempty
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:9]: rm

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -3,6 +3,11 @@
 
 Text retrieval with BM25
 """
+
+
+from docutils.nodes import document
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:8]: rm

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -18,6 +23,21 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str) or not text:
+        return None
+    result = ""
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:28]: try not to create a new variable

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -18,6 +23,21 @@ def tokenize(text: str) -> list[str] | None:
...
+    for elem in text.lower():
+        if elem.isalpha() is True:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:30]: `is True` is not required

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -18,6 +23,21 @@ def tokenize(text: str) -> list[str] | None:
...
+            result += elem
+            continue
+        result += " "
+    tokens1 = result.split(" ")
+    tokens2 = []
+    for elem in tokens1:
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:36]: is this loop really necessary?

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -77,6 +119,14 @@ def calculate_idf(vocab: list[str], documents: list[list[str]]) -> dict[str, flo
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(vocab, list) or not all(isinstance(item, str) for item in vocab) or not isinstance(documents, list) or not all(isinstance(docum, list) for docum in documents) or not all(isinstance(item, str) for docum in documents for item in docum) or not vocab or not documents:
+        return None
+    idf_result = {}
+    doc_num = len(documents)
+    for word in vocab:
+        words_count = sum(1 for doc in documents if word in doc)
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:127]: great using `sum()`!

**Review by Vasilisa-Blyudova — COMMENTED** [review]
Great work! Try to implement a work with a score of 10. 
Good luck

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:8]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:9]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:28]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 9 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:30]: Done

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
Comment (QuietHellsPage) [reply lab_2_retrieval_w_bm25/main.py:36]: Done

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -3,7 +3,8 @@
 
 Text retrieval with BM25
 """
-# pylint:disable=too-many-arguments, unused-argument
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:6]: revert

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -18,6 +19,10 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str) or not text:
+        return None
+    return [word for word in ''.join(char if char.isalpha()
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:24]: don't understand, simplify

File: lab_2_retrieval_w_bm25/main.py [hunk 14]
```diff
@@ -138,72 +210,90 @@ def rank_documents(
...
+        return None
+    clear_query = remove_stopwords(tok_query, stopwords)
+    if not isinstance(clear_query, list):
+        return None
+    result = [(i, round(sum(dictionary.get(word, 0) for word in clear_query), 4))
+              for i, dictionary in enumerate(indexes)]
+    return sorted(result, reverse=True, key=lambda res: res[1])
+
+
+# def calculate_bm25_with_cutoff(
```
Comment (Vasilisa-Blyudova) [inline lab_2_retrieval_w_bm25/main.py:233]: revert

**Pull request closed without merging**